lib/executor/variable_arrival_rate.go

```diff
@@ -178,9 +178,24 @@ func (varc VariableArrivalRateConfig) getPlannedRateChanges(segment *lib.Executi
 	// the rational value for scale(20%, 1/sec) is 0.2/sec, or rather 1/5sec...
 	currentRate := getScaledArrivalRate(segment, varc.StartRate.Int64, timeUnit)
 
-	rateChanges := []rateChange{}
+	var size int
+	for i, stage := range varc.Stages {
+		if stage.Duration.Duration == 0 &&
+			(i == 0 ||
+				(i >= 2 && varc.Stages[i-1].Target.Int64 != stage.Target.Int64 &&
+					varc.Stages[i-1].Target.Int64 == varc.Stages[i-2].Target.Int64)) {
+			size++
+		} else if i == 0 || varc.Stages[i-1].Target.Int64 != stage.Target.Int64 {
+			size += int(time.Duration(stage.Duration.Duration) / minIntervalBetweenRateAdjustments)
+		}
+	}
+
+	rateChanges := make([]rateChange, 0, size)
```

Comment:
This seems a hugely complicated piece of code for what (I think?) is essentially a very minor and ultra premature optimization...

Reply:
I do agree that it grew much harrier than it was originally (I think I've squashed those commits) and the performance benefit turned out to not be as big as expected ;( 
Do you want me to drop it ? 
cc @imiric , @cuonglm

Reply:
Yeah, I vote :+1:  for dropping this. Now that I look at it again, I think it doesn't have a significant performance benefit, since you'd probably be saving some memory, but at the expense of actually calculating things almost twice - once here and once below...

Reply:
I have added the benchmark results ;) https://github.com/loadimpact/k6/pull/1117/commits/93ff4da57cfb0663aee61a83a4b5e9017b9e8031

Reply:
Didn't investigate the benchmark very thoroughly, but  I find it a bit strange that this function takes up to a megabyte of RAM :confused: to execute...

Reply:
This function is already tricky to follow, so if this isn't a critical optimization, I also vote :+1: for removing this.

Reply:
look at 143b9b226c3731d8780b3fe21aa15844387b7ee5 , I do calculate the ramping part of the benchmark to be around 28800 bytes for the structs in the slice, so with the same amount for the ramp down .. around 60k ?

Reply:
I'm confused, the `old bytes` / `new bytes` I see in https://github.com/loadimpact/k6/commit/93ff4da57cfb0663aee61a83a4b5e9017b9e8031 are all ~1MB, which seems very excessive for this method...

Reply:
one of the reasons why I took some time optimizing this was because ... it was creating way too many objects and as is shown in 143b9b226c3731d8780b3fe21aa15844387b7ee5 , not creating 3 `big.Rat` reduces the usage by a 34% ... 

There are more `big.Rat`s being created constantly (because calculations) and for the 5 minutes ramp up you need to calculate `5 * 60 * 1000 / 250 = 1200` steps .. so 2400 steps with the ramp down (roughly). So 2400 we need to go through this cycle of calculating the steps... In all of those cases we create `big.Rat`, which is actually 2 `big.Int`'s,  which is actually a bool and a [nat](https://golang.org/src/math/big/nat.go) (which is not constant sized :( ) . All of this adds up if we constantly have to allocate new ones for each of the steps .. which is what was happening
The reusing of the three `big.Rat`s roughly calculates to `354733 (saved) / 2400(steps) = 147.80541666666667` bytes of saved space per iterations (roughly). 
I did all the optimizations with the help of profiling the benchmark which basically told me that we primarily allocated `big.Rat/big.Int` all the time ... we still do .. it's just now inside of the calculations and I can't reduce this ... I think ...

Reply:
Dropped it from the changes and putted in another branch for the history. Below is the benchstat between the last changes here and this :
```
    name                     old time/op    new time/op    delta
    GetPlannedRateChanges-4    1.39ms ±11%    1.36ms ± 4%     ~     (p=1.000 n=5+5)

    name                     old alloc/op   new alloc/op   delta
    GetPlannedRateChanges-4     863kB ± 0%     675kB ± 0%  -21.83%  (p=0.008 n=5+5)

    name                     old allocs/op  new allocs/op  delta
    GetPlannedRateChanges-4     37.2k ± 0%     37.2k ± 0%   -0.04%  (p=0.008 n=5+5)

```

Reply:
Ah, OK, that makes sense. Still not sure why https://github.com/loadimpact/k6/commit/93ff4da57cfb0663aee61a83a4b5e9017b9e8031 drops the memory usage by ~250kb, given that it doesn't affect any `big.Rat` allocation :confused: It just properly sizes `rateChanges`, which is a simple slice of `{time.Duration, types.NullDuration} `, basically ~20 bytes :confused:

Reply:
btw I am wondering wif the `types.NullDuration` can't be just `time.Duration` ... but I will need to look at the code ... as this makes everything around this funkier to write ...

Reply:
IIRC, I intentionally used a `NullDuration`, to make it apparent that the value could be `0` (given that we can't pass that zero value to `time.NewTicker()` - see the comments in `type rateChange struct {}`

Reply:
comparing the memory profile using the current tip as the `base` (also some changes so that the lines stay the same):
```
      flat  flat%   sum%        cum   cum%
 -195.48MB 12.19% 12.19%  -198.98MB 12.41%  github.com/loadimpact/k6/lib/executor.VariableArrivalRateConfig.getPlannedRateChanges
     -17MB  1.06% 13.25%      -17MB  1.06%  math/big.nat.make
   13.50MB  0.84% 12.41%       -2MB  0.12%  math/big.scaleDenom
    1.07MB 0.067% 12.34%     1.07MB 0.067%  archive/zip.readDirectoryEnd
         0     0% 12.34%     1.07MB 0.067%  archive/zip.(*Reader).init
         0     0% 12.34%     1.07MB 0.067%  archive/zip.NewReader
```
```
(pprof) list getPlannedRateChanges
Total: 1.57GB
ROUTINE ======================== github.com/loadimpact/k6/lib/executor.VariableArrivalRateConfig.getPlannedRateChanges in /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/lib/executor/var$
able_arrival_rate.go
 -195.48MB  -198.98MB (flat, cum) 12.41% of Total
         .          .    178:   // numbers for scaling, then the instance executing the first segment won't
         .          .    179:   // ever do even a single request, since scale(20%, 1) would be 0, whereas
         .          .    180:   // the rational value for scale(20%, 1/sec) is 0.2/sec, or rather 1/5sec...
         .          .    181:   currentRate := getScaledArrivalRate(segment, varc.StartRate.Int64, timeUnit)
         .          .    182:
   69.15MB    69.15MB    183:   rateChanges := make([]rateChange, 0, varc.calcSize())
         .          .    184:   timeFromStart := time.Duration(0)
         .          .    185:
         .          .    186:   var tArrivalRate = new(big.Rat)
         .          .    187:   var tArrivalRateStep = new(big.Rat)
         .          .    188:   var stepCoef = new(big.Rat)
         .          .    189:   for _, stage := range varc.Stages {
         .          .    190:           stageTargetRate := getScaledArrivalRate(segment, stage.Target.Int64, timeUnit)
         .          .    191:           stageDuration := time.Duration(stage.Duration.Duration)
         .          .    192:
         .        1MB    193:           if currentRate.Cmp(stageTargetRate) == 0 {
         .          .    194:                   // We don't have to do anything but update the time offset
         .          .    195:                   // if the rate wasn't changed in this stage
         .          .    196:                   timeFromStart += stageDuration
         .          .    197:                   continue
/// a lot of nothing 
         .          .    218:           stepNumber := (stageDuration / minIntervalBetweenRateAdjustments)
         .          .    219:           if stepNumber > 1 {
         .        -8B    220:                   rateDiff := new(big.Rat).Sub(stageTargetRate, currentRate)
         .          .    221:                   stepInterval := stageDuration / stepNumber
         .          .    222:                   for t := stepInterval; ; t += stepInterval {
         .          .    223:                           if stageDuration-t < minIntervalBetweenRateAdjustments {
         .          .    224:                                   break
         .          .    225:                           }
         .          .    226:
         .   -16.50MB    227:                           tArrivalRate.Add(
         .          .    228:                                   currentRate,
         .        1MB    229:                                   tArrivalRateStep.Mul(
         .          .    230:                                           rateDiff,
         .        8MB    231:                                           stepCoef.SetFrac64(int64(t), int64(stageDuration)),
         .          .    232:                                   ),
         .          .    233:                           )
         .          .    234:
 -264.62MB  -264.62MB    235:                           rateChanges = append(rateChanges, rateChange{
         .          .    236:                                   timeOffset:   timeFromStart + t,
         .        3MB    237:                                   tickerPeriod: getTickerPeriod(tArrivalRate),
         .          .    238:                           })
         .          .    239:                   }
         .          .    240:           }
         .          .    241:           timeFromStart += stageDuration
         .          .    242:           rateChanges = append(rateChanges, rateChange{
```

Reply:
both benchmarks ran 1000 times so ... all the numbers should be divided by 1000 so ... around 260kb saved by ... probably not having to increase the underlying slice a bunch of times and than throwing it away as it still needs to grow ...

Reply:
> IIRC, I intentionally used a `NullDuration`, to make it apparent that the value could be `0` (given that we can't pass that zero value to `time.NewTicker()` - see the comments in `type rateChange struct {}`

I saw the comment it's just that `NullDuration` adds a boolean and in the same vein as this whole thread - this literally means that if we are not using it we will save 33.3% of the memory used for [rateChange struct](http://golang-sizeof.tips/?t=Ly8gU2FtcGxlIGNvZGUKc3RydWN0IHsKCWEgaW50NjQKCWIgYm9vbAp9Cg==) (time.Duraiton is an int64 and null.Duraiton is time.Duration with a bool). 
Not a huge win but given that a similar thing can be accomplished with either a pointer to time.Duration or ... just you know knowing that it can't be 0  and checking for it where appropriate ...